src/file-table-engine/src/table/format.rs

```diff
@@ -85,6 +86,19 @@ fn build_json_opener(
     ))
 }
 
+fn build_orc_opener(file_schema: Arc<ArrowSchema>, config: &ScanPlanConfig) -> Result<OrcOpener> {
+    let projected_schema = if let Some(projection) = config.projection {
+        Arc::new(
+            file_schema
+                .project(projection)
+                .context(error::ProjectSchemaSnafu)?,
+        )
+    } else {
+        file_schema
+    };
+    Ok(OrcOpener::new(config.store.clone(), projected_schema))
```

Comment:
Can you add a projection test like: 
https://github.com/GreptimeTeam/greptimedb/blob/b9bee9128de2c580bf2ec1637813eef9b51ad490/tests-integration/src/tests/instance_test.rs#L518-L585


You will run into a bug because it doesn't align the stream's origin schema to the projected schema(output schema).

https://github.com/GreptimeTeam/greptimedb/blob/b9bee9128de2c580bf2ec1637813eef9b51ad490/src/common/datasource/src/file_format/orc.rs#L89-L111

Considers to change it like the following: (remember to remove the unwrap)
```rust
        let batch = batch.map(|b| {
            b.and_then(|b| {
                let mut columns = Vec::with_capacity(self.output_schema.fields().len());

                for field in self.output_schema.fields().into_iter() {
                    let column = b.column_by_name(field.name()).unwrap();
                    if column.data_type() != field.data_type() {
                        let output = cast(&column, field.data_type())?;
                        columns.push(output)
                    } else {
                        columns.push(column.clone())
                    }
                }
                let record_batch = DfRecordBatch::try_new(self.output_schema.clone(), columns)?;

                Ok(record_batch)
            })
        });
```

Reply:
Considering the performance issues, let's add a new struct for stream projection. Maybe we can utilize the pre-computed to avoid unnecessary computing.

Reply:
> Can you add a projection test like:
could I move test.orc(including various type) to `tests/data`? And test orc_opener with basic.orc which only has two types int64 and utf8?

Reply:
How about just the following?
```rust
    let filepath = get_data_dir("../src/common/datasource/tests/orc/test.orc")
        .canonicalize()
        .unwrap()
        .display()
        .to_string();
```